### PR TITLE
ENH: Implement fastpath for masked indexes in join

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

As of right now,  masked indexes go through object even though we only use the join functions when we are monotonic, e.g. no NA present. We can simply use the numpy array to perform the join.

```
       before           after         ratio
     [026a83e0]       [d804bced]
-     5.82±0.02ms      1.38±0.02ms     0.24  index_object.SetOperations.time_operation('monotonic', 'ea_int', 'intersection')
-     4.08±0.03ms         452±10μs     0.11  index_object.SetOperations.time_operation('monotonic', 'ea_int', 'union')
```

This is related to #49420, will make it a bit easier to review since it solves the join changes

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4799,11 +4803,23 @@ def _get_engine_target(self) -> ArrayLike:
             return self._values.astype(object)
         return vals
 
+    def _get_join_target(self) -> ArrayLike:
+        """
+        Get the ndarray or ExtensionArray that we can pass to the join
+        functions.
+        """
+        if isinstance(self._values, BaseMaskedArray):
+            # This is only used if our array is monotonic, so no NAs present
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:4812]: Where is monoticity checked of `self._values._data`?

Comment (phofl) [reply pandas/core/indexes/base.py:4812]: Using those functions is guarded by those checks. The actual join implementations here only work with monotonic data. As soon as we reach ``_inner_indexer`` and others we have to be monotonic

Comment (mroeschke) [reply pandas/core/indexes/base.py:4812]: Ah okay cool

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**